### PR TITLE
[MRG+1] Parse robots.txt content as native str

### LINKED ISSUE #2373: Wikipedia robots.txt raises exceptions
I'm scraping a page which in turn links to wikipedia.

But the wikipedia robots.txt is creating some errors/exceptions as below.

Python 2.7.12
Scrapy 1.2.1

```
2016-11-02 13:13:18 [scrapy] DEBUG: Crawled (200) <GET https://en.wikipedia.org/robots.txt> (referer: None)
2016-11-02 13:13:18 [py.warnings] WARNING: C:\Python27\lib\urllib.py:1303: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  return ''.join(map(quoter, s))

2016-11-02 13:13:18 [scrapy] ERROR: Error downloading <GET http://en.wikipedia.org/robots.txt>: u'\xd8'
Traceback (most recent call last):
  File "C:\Python27\lib\site-packages\twisted\internet\defer.py", line 587, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "C:\Python27\lib\site-packages\scrapy\downloadermiddlewares\robotstxt.py", line 97, in _parse_robots
    rp.parse(body.splitlines())
  File "C:\Python27\lib\robotparser.py", line 120, in parse
    entry.rulelines.append(RuleLine(line[1], False))
  File "C:\Python27\lib\robotparser.py", line 174, in __init__
    self.path = urllib.quote(path)
  File "C:\Python27\lib\urllib.py", line 1303, in quote
    return ''.join(map(quoter, s))
KeyError: u'\xd8'
```

### PR DESCRIPTION
**Pull request by redapple**
Fixes #2373 

Simpler fix than https://github.com/scrapy/scrapy/pull/2385

### CODE REVIEW

File: scrapy/downloadermiddlewares/robotstxt.py [hunk 1]
```diff
@@ -94,7 +95,9 @@ def _parse_robots(self, response, netloc):
                 # Running rp.parse() will set rp state from
                 # 'disallow all' to 'allow any'.
                 pass
-        rp.parse(body.splitlines())
+        # stdlib's robotparser expects native 'str' ;
+        # with unicode input, non-ASCII encoded bytes decoding fails in Python2
+        rp.parse(to_native_str(body).splitlines())
```
Comment (kmike) [inline scrapy/downloadermiddlewares/robotstxt.py]: Nice! 

This uses utf-8 in Python 2, even if original page encoding is different, right? Is this correct? Likely it is, but I don't understand the fix well enough.

**Comment by kmike** [comment]
It could be a good idea to make a robots.txt crawl at some point - get a dataset of e.g. 100000 robots.txt files from Alexa top 1M list and check which problems Scrapy has parsing them.

**Comment by redapple** [comment]
Very good point @kmike.
This patch introduces a regression for some robots.txt in the wild, eg. http://www.quikr.com/robots.txt

```
2016-11-15 10:59:28 [scrapy] ERROR: Error downloading <GET http://www.quikr.com>
Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapydev/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 587, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/paul/src/scrapy/scrapy/downloadermiddlewares/robotstxt.py", line 44, in process_request_2
    if rp is not None and not rp.can_fetch(self._useragent, request.url):
  File "/usr/lib/python2.7/robotparser.py", line 153, in can_fetch
    if entry.applies_to(useragent):
  File "/usr/lib/python2.7/robotparser.py", line 207, in applies_to
    if agent in useragent:
UnicodeDecodeError: 'ascii' codec can't decode byte 0xef in position 4: ordinal not in range(128)
```

**Comment by redapple** [comment]
A couple of errors I saw as well (not directly related to robotparser per se though):

```
2016-11-15 11:29:36 [scrapy] ERROR: Error downloading <GET http://www.huffingtonpost.kr>
Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapydev/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1185, in _inlineCallbacks
    result = g.send(result)
  File "/home/paul/src/scrapy/scrapy/core/downloader/middleware.py", line 53, in process_response
    spider=spider)
  File "/home/paul/src/scrapy/scrapy/downloadermiddlewares/httpcompression.py", line 30, in process_response
    decoded_body = self._decode(response.body, encoding.lower())
  File "/home/paul/src/scrapy/scrapy/downloadermiddlewares/httpcompression.py", line 46, in _decode
    body = gunzip(body)
  File "/home/paul/src/scrapy/scrapy/utils/gz.py", line 37, in gunzip
    chunk = read1(f, 8196)
  File "/home/paul/src/scrapy/scrapy/utils/gz.py", line 21, in read1
    return gzf.read(size)
  File "/usr/lib/python2.7/gzip.py", line 268, in read
    self._read(readsize)
  File "/usr/lib/python2.7/gzip.py", line 303, in _read
    self._read_gzip_header()
  File "/usr/lib/python2.7/gzip.py", line 197, in _read_gzip_header
    raise IOError, 'Not a gzipped file'
IOError: Not a gzipped file
```

```
2016-11-15 11:24:27 [scrapy] ERROR: Error downloading <GET http://www.huffpostarabi.com/robots.txt>: Not a gzipped file
Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapydev/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1185, in _inlineCallbacks
    result = g.send(result)
  File "/home/paul/src/scrapy/scrapy/core/downloader/middleware.py", line 53, in process_response
    spider=spider)
  File "/home/paul/src/scrapy/scrapy/downloadermiddlewares/httpcompression.py", line 30, in process_response
    decoded_body = self._decode(response.body, encoding.lower())
  File "/home/paul/src/scrapy/scrapy/downloadermiddlewares/httpcompression.py", line 46, in _decode
    body = gunzip(body)
  File "/home/paul/src/scrapy/scrapy/utils/gz.py", line 37, in gunzip
    chunk = read1(f, 8196)
  File "/home/paul/src/scrapy/scrapy/utils/gz.py", line 21, in read1
    return gzf.read(size)
  File "/usr/lib/python2.7/gzip.py", line 268, in read
    self._read(readsize)
  File "/usr/lib/python2.7/gzip.py", line 303, in _read
    self._read_gzip_header()
  File "/usr/lib/python2.7/gzip.py", line 197, in _read_gzip_header
    raise IOError, 'Not a gzipped file'
IOError: Not a gzipped file
```

**Comment by redapple** [comment]
Regarding http://www.huffpostarabi.com/, it looks like the webserver is saying it's gzip encoding the response, but instead serves a plain XML document along with a 403 error:

```
$ scrapy shell http://www.huffpostarabi.com/ -s ROBOTSTXT_OBEY=1
2016-11-15 16:47:03 [scrapy] INFO: Scrapy 1.2.1 started (bot: scrapybot)
(...)
2016-11-15 16:47:04 [scrapy] INFO: Spider opened
403
{'Content-Encoding': ['gzip'], 'X-Ec-Security-Audit': ['403'], 'Expires': ['Tue, 15 Nov 2016 15:57:04 GMT'], 'Server': ['ECD (ory/4393)'], 'Cache-Control': ['max-age=600'], 'Date': ['Tue, 15 Nov 2016 15:47:04 GMT'], 'Content-Type': ['text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8']}
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
         "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
	<head>
		<title>403 - Forbidden</title>
	</head>
	<body>
		<h1>403 - Forbidden</h1>
	</body>
</html>

2016-11-15 16:47:04 [scrapy] ERROR: Error downloading <GET http://www.huffpostarabi.com/robots.txt>: Not a gzipped file
Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapydev/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1185, in _inlineCallbacks
    result = g.send(result)
  File "/home/paul/src/scrapy/scrapy/core/downloader/middleware.py", line 53, in process_response
    spider=spider)
  File "/home/paul/src/scrapy/scrapy/downloadermiddlewares/httpcompression.py", line 33, in process_response
    decoded_body = self._decode(response.body, encoding.lower())
  File "/home/paul/src/scrapy/scrapy/downloadermiddlewares/httpcompression.py", line 49, in _decode
    body = gunzip(body)
  File "/home/paul/src/scrapy/scrapy/utils/gz.py", line 37, in gunzip
    chunk = read1(f, 8196)
  File "/home/paul/src/scrapy/scrapy/utils/gz.py", line 21, in read1
    return gzf.read(size)
  File "/usr/lib/python2.7/gzip.py", line 268, in read
    self._read(readsize)
  File "/usr/lib/python2.7/gzip.py", line 303, in _read
    self._read_gzip_header()
  File "/usr/lib/python2.7/gzip.py", line 197, in _read_gzip_header
    raise IOError, 'Not a gzipped file'
IOError: Not a gzipped file
403
{'Content-Encoding': ['gzip'], 'X-Ec-Security-Audit': ['403'], 'Expires': ['Tue, 15 Nov 2016 15:47:34 GMT'], 'Server': ['ECD (ory/43A1)'], 'Cache-Control': ['max-age=30'], 'Date': ['Tue, 15 Nov 2016 15:47:04 GMT'], 'Content-Type': ['text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8']}
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
         "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
	<head>
		<title>403 - Forbidden</title>
	</head>
	<body>
		<h1>403 - Forbidden</h1>
	</body>
</html>

Traceback (most recent call last):
  File "/home/paul/.virtualenvs/scrapydev/bin/scrapy", line 11, in <module>
    load_entry_point('Scrapy', 'console_scripts', 'scrapy')()
  File "/home/paul/src/scrapy/scrapy/cmdline.py", line 142, in execute
    _run_print_help(parser, _run_command, cmd, args, opts)
  File "/home/paul/src/scrapy/scrapy/cmdline.py", line 88, in _run_print_help
    func(*a, **kw)
  File "/home/paul/src/scrapy/scrapy/cmdline.py", line 149, in _run_command
    cmd.run(args, opts)
  File "/home/paul/src/scrapy/scrapy/commands/shell.py", line 71, in run
    shell.start(url=url)
  File "/home/paul/src/scrapy/scrapy/shell.py", line 47, in start
    self.fetch(url, spider)
  File "/home/paul/src/scrapy/scrapy/shell.py", line 112, in fetch
    reactor, self._schedule, request, spider)
  File "/home/paul/.virtualenvs/scrapydev/local/lib/python2.7/site-packages/twisted/internet/threads.py", line 122, in blockingCallFromThread
    result.raiseException()
  File "<string>", line 2, in raiseException
IOError: Not a gzipped file
```

We can add a try/except in http compression middleware perhaps. Not sure.

**Comment by kmike** [comment]
Browser seems to handle this website fine. I wonder if we need to dive deep and implement something like https://mimesniff.spec.whatwg.org/ eventually.

**Comment by redapple** [comment]
https://github.com/scrapy/scrapy/pull/2388/commits/6cd35c77da6601f218b5042f4b9a9919e642e15c fixes the error in https://github.com/scrapy/scrapy/pull/2388#issuecomment-260603669

Without the code patch, the tests fail with
```
/home/paul/src/scrapy/scrapy/downloadermiddlewares/robotstxt.py:46: in process_request_2
    self._useragent, request.url):
/usr/lib/python2.7/robotparser.py:153: in can_fetch
    if entry.applies_to(useragent):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <robotparser.Entry instance at 0x7fb88c64f0e0>, useragent = 'scrapy'

    def applies_to(self, useragent):
        """check if this entry applies to the specified agent"""
        # split the name token and make it lower case
        useragent = useragent.split("/")[0].lower()
        for agent in self.useragents:
            if agent == '*':
                # we have the catch-all agent
                return True
            agent = agent.lower()
>           if agent in useragent:
E           UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 4: ordinal not in range(128)

/usr/lib/python2.7/robotparser.py:207: UnicodeDecodeError
```

**Comment by redapple** [comment]
@kmike , it looks like http://www.huffpostarabi.com/robots.txt responses depend on User-Agent string. With "scrapy" or my Chrome's value, I can't reproduce `IOError: Not a gzipped file`

**Comment by redapple** [comment]
@kmike , MIME sniffing looks like an interesting addition for w3lib

**Comment by kmike** [comment]
Looks good 👍

**Pull request merged**